crates/primitives/src/precompile.rs

```diff
@@ -6,7 +6,32 @@ use std::{boxed::Box, string::String, sync::Arc};
 /// A precompile operation result.
 ///
 /// Returns either `Ok((gas_used, return_bytes))` or `Err(error)`.
-pub type PrecompileResult = Result<(u64, Bytes), PrecompileError>;
+pub enum PrecompileResult {
+    Ok { gas_used: u64, output: Bytes },
+    Error { error_type: PrecompileError },
+    FatalError { msg: String },
+}
```

Comment:
Maybe we should structure this differently and still use the `Result` at the end it enables `?`.

So this would mean having; `Result<PrecompileOutput, PrecompileErrors>`

Where:
```
enum PrecompileErrors {
    Error(PrecompileError),
    Fatal {
        msg: String
    },
}
```

And maybe where we are at it add `PrecompileOutput`

Reply:
okay working on it!